### PR TITLE
blk/pmem: Add the devdax support.

### PR DESCRIPTION
**Pull request by optimistyzy**
The purpose is to make the pmem device usage more flexible
than the current solution. And prepare for the potential
offloading by hardware engine later.

Signed-off-by: Ziye Yang <ziye.yang@intel.com>

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/blk/pmem/PMEMDevice.cc [hunk 1]
```diff
@@ -82,13 +144,21 @@ int PMEMDevice::open(const std::string& p)
     goto out_fail;
   }
 
-  size_t map_len;
-  addr = (char *)pmem_map_file(path.c_str(), 0, PMEM_FILE_EXCL, O_RDWR, &map_len, NULL);
-  if (addr == NULL) {
-    derr << __func__ << " pmem_map_file failed: " << pmem_errormsg() << dendl;
-    goto out_fail;
+  if (devdax_device) {
+    addr = (char *) mmap(NULL, size, PROT_READ | PROT_WRITE, MAP_SHARED, fd, 0);
```
Comment (majianpeng) [inline src/blk/pmem/PMEMDevice.cc:148]: we can still use pmem_map_file for devdax.

File: src/blk/pmem/PMEMDevice.cc [hunk 2]
```diff
@@ -168,7 +243,22 @@ bool PMEMDevice::support(const std::string &path)
...
+  local_fd = ::open(path.c_str(), O_RDWR);
+  if (local_fd < 0) {
+    return false;
+  }
+
+  r = pmem_check_file_type(local_fd, path.c_str(), NULL);
+  VOID_TEMP_FAILURE_RETRY(::close(local_fd));
+  if (r == 0) {
+    return true;
+  }
```
Comment (majianpeng) [inline src/blk/pmem/PMEMDevice.cc:259]: why add those code?

File: src/blk/pmem/PMEMDevice.h [hunk 3]
```diff
@@ -40,7 +41,7 @@ class PMEMDevice : public BlockDevice {
 public:
   PMEMDevice(CephContext *cct, aio_callback_t cb, void *cbpriv);
 
-
+  bool supported_bdev_label() override { return !devdax_device; };
```
Comment (majianpeng) [inline src/blk/pmem/PMEMDevice.h:44]: Where is this function called?

File: src/blk/pmem/PMEMDevice.cc [hunk 4]
```diff
@@ -18,6 +18,13 @@
 #include <stdlib.h>
 #include <sys/types.h>
 #include <sys/stat.h>
+#include <sys/sysmacros.h>
+#include <stdio.h>
+#include <errno.h>
+#include <sys/stat.h>
+#include <sys/sysmacros.h>
```
Comment (tchaikov) [inline src/blk/pmem/PMEMDevice.cc:25]: this file gets included twice.

Comment (optimistyzy) [reply src/blk/pmem/PMEMDevice.cc:25]: removed

File: src/blk/pmem/PMEMDevice.cc (hunk 2 above)
Comment (optimistyzy) [reply src/blk/pmem/PMEMDevice.cc:263]: checking whether it is a devdax file. then use different open flags

File: src/blk/pmem/PMEMDevice.h (hunk 3 above)
Comment (optimistyzy) [reply src/blk/pmem/PMEMDevice.h:44]: called in Bluefs.cc or BlueStore.cc. For using /dev/dax, it is not a block device. So we cannot allow normal block operations, which is same to use SPDK to drive nvme devices.

**Comment by optimistyzy** [comment]
@tchaikov updated version. With this patch,  we can support using PMEM char device.

File: src/blk/pmem/PMEMDevice.cc [hunk 5]
```diff
@@ -54,10 +59,61 @@ int PMEMDevice::_lock()
   return 0;
 }
 
+static int pmem_check_file_type(int fd, const char *pmem_file, uint64_t *total_size)
+{
+  int rc = 0;
+  struct stat file_stat;
+  char spath[PATH_MAX];
+  char npath[PATH_MAX];
+  char *real_path, *base_name;
```
Comment (tchaikov) [inline src/blk/pmem/PMEMDevice.cc:68]: please define variable close to where it is used for the first time for better readability.

File: src/blk/pmem/PMEMDevice.cc [hunk 6]
```diff
@@ -54,10 +59,61 @@ int PMEMDevice::_lock()
...
+  }
+
+  if ((file_stat.st_mode & S_IFCHR) != S_IFCHR) {
+    return -1;
+  }
+
+  snprintf(spath, PATH_MAX, "/sys/dev/char/%d:%d/subsystem",
+           major(file_stat.st_rdev), minor(file_stat.st_rdev));
+
+  real_path = realpath(spath, npath);
```
Comment (tchaikov) [inline src/blk/pmem/PMEMDevice.cc:83]:
```c++
namespace fs = std::filesystem
if (!fs::is_character_file(pmem_file)) {
  return -EINVAL;
}
struct stat file_stat;
if (::fstat(fd, &file_stat)) {
  return -EINVAL;
}
auto char_dir = fmt::format("/sys/dev/char/{}:{}",
                            major(file_stat.st_rdev), 
                            minor(file_stat.st_rdev));
if (auto subsys_path = char_path / "subsystem";
    fs::read_symlink(subsys_path).filename().string() != "dax") {
  return -EINVAL;
}
if (total_size == nullptr) {
  return 0;
}
if (std::ifstream size_file(char_path / "size"); size_file)
  size_file >> *total_size;
  return size_file ? 0 : -EINVAL;
} else {
  return -EINVAL;
}
```

i'd suggest rewrite this part usinf `std::filesystem()` for better readability.

File: src/blk/pmem/PMEMDevice.h [hunk 7]
```diff
@@ -30,6 +30,7 @@ class PMEMDevice : public BlockDevice {
   int fd;
   char *addr; //the address of mmap
   std::string path;
+  bool devdax_device;
```
Comment (tchaikov) [inline src/blk/pmem/PMEMDevice.h:33]:
```suggestion
  bool devdax_device = false;
```

File: src/blk/pmem/PMEMDevice.cc [hunk 8]
```diff
@@ -168,13 +243,30 @@ bool PMEMDevice::support(const std::string &path)
...
+
+  local_fd = ::open(path.c_str(), O_RDWR);
+  if (local_fd < 0) {
+    return false;
+  }
+
+  r = pmem_check_file_type(local_fd, path.c_str(), NULL);
+  VOID_TEMP_FAILURE_RETRY(::close(local_fd));
+  if (r == 0) {
+      flags = 0;
```
Comment (tchaikov) [inline src/blk/pmem/PMEMDevice.cc:259]: wrong indent.

File: src/blk/pmem/PMEMDevice.cc [hunk 9]
```diff
@@ -168,13 +243,30 @@ bool PMEMDevice::support(const std::string &path)
 {
   int is_pmem = 0;
   size_t map_len = 0;
-  void *addr = pmem_map_file(path.c_str(), 0, PMEM_FILE_EXCL, O_RDONLY, &map_len, &is_pmem);
+  void *addr;
```
Comment (tchaikov) [inline src/blk/pmem/PMEMDevice.cc:246]: please define variable close to where it is used for the first time for better readability.

File: src/blk/pmem/PMEMDevice.h (hunk 7 above)
Comment (optimistyzy) [reply src/blk/pmem/PMEMDevice.h:33]: ack

File: src/blk/pmem/PMEMDevice.cc (hunk 5 above)
Comment (optimistyzy) [reply src/blk/pmem/PMEMDevice.cc:68]: ack

File: src/blk/pmem/PMEMDevice.cc (hunk 6 above)
Comment (optimistyzy) [reply src/blk/pmem/PMEMDevice.cc:83]: Can we revise it later?

File: src/blk/pmem/PMEMDevice.cc (hunk 9 above)
Comment (optimistyzy) [reply src/blk/pmem/PMEMDevice.cc:246]: ack

File: src/blk/pmem/PMEMDevice.cc (hunk 8 above)
Comment (optimistyzy) [reply src/blk/pmem/PMEMDevice.cc:259]: ack

File: src/blk/pmem/PMEMDevice.cc [hunk 10]
```diff
@@ -35,7 +40,7 @@
 
 PMEMDevice::PMEMDevice(CephContext *cct, aio_callback_t cb, void *cbpriv)
   : BlockDevice(cct, cb, cbpriv),
-    fd(-1), addr(0),
+    fd(-1), addr(0), devdax_device(false),
```
Comment (tchaikov) [inline src/blk/pmem/PMEMDevice.cc:43]: now that this variable is initialized in-class, could just drop this line.

File: src/blk/pmem/PMEMDevice.cc [hunk 11]
```diff
@@ -54,10 +59,59 @@ int PMEMDevice::_lock()
   return 0;
 }
 
+static int pmem_check_file_type(int fd, const char *pmem_file, uint64_t *total_size)
+{
+  int rc = 0;
+  struct stat file_stat;
+  FILE *sfile;
```
Comment (tchaikov) [inline src/blk/pmem/PMEMDevice.cc:66]: please define a variable closer to where it is used for better readability.

File: src/blk/pmem/PMEMDevice.cc [hunk 12]
```diff
@@ -54,10 +59,59 @@ int PMEMDevice::_lock()
...
+    return -1;
+  }
+
+  snprintf(spath, PATH_MAX, "/sys/dev/char/%d:%d/size", major(file_stat.st_rdev), minor(file_stat.st_rdev));
+  sfile = fopen(spath, "r");
+  if (!sfile) {
+      return -1;
+  }
+
+  if (total_size != NULL) {
```
Comment (tchaikov) [inline src/blk/pmem/PMEMDevice.cc:98]:
```suggestion
  if (total_size != nullptr) {
```

more a là mode this way.

File: src/blk/pmem/PMEMDevice.cc [hunk 13]
```diff
@@ -54,10 +59,59 @@ int PMEMDevice::_lock()
...
+    rc = fscanf(sfile, "%lu", total_size);
+    if (rc < 0) {
+      rc = -1;
```
Comment (tchaikov) [inline src/blk/pmem/PMEMDevice.cc:101]: this variable is assigned, but never read.

File: src/blk/pmem/PMEMDevice.cc [hunk 14]
```diff
@@ -54,10 +59,59 @@ int PMEMDevice::_lock()
...
+
+  fclose(sfile);
+  return 0;
+}
+
 int PMEMDevice::open(const std::string& p)
 {
   path = p;
   int r = 0;
+  int flags = 0;
```
Comment (tchaikov) [inline src/blk/pmem/PMEMDevice.cc:113]: please define a variable closer to where it is used for better readability.

File: src/blk/pmem/PMEMDevice.h (hunk 3 above)
Comment (tchaikov) [reply src/blk/pmem/PMEMDevice.h:44]: might want to remove the trailing `;`.

File: src/blk/pmem/PMEMDevice.cc (hunk 10 above)
Comment (optimistyzy) [reply src/blk/pmem/PMEMDevice.cc:43]: ack

File: src/blk/pmem/PMEMDevice.cc (hunk 11 above)
Comment (optimistyzy) [reply src/blk/pmem/PMEMDevice.cc:66]: ack

File: src/blk/pmem/PMEMDevice.cc (hunk 12 above)
Comment (optimistyzy) [reply src/blk/pmem/PMEMDevice.cc:98]: ack

File: src/blk/pmem/PMEMDevice.cc (hunk 14 above)
Comment (optimistyzy) [reply src/blk/pmem/PMEMDevice.cc:113]: ack

File: src/blk/pmem/PMEMDevice.cc (hunk 13 above)
Comment (optimistyzy) [reply src/blk/pmem/PMEMDevice.cc:101]: ack

File: src/blk/pmem/PMEMDevice.h (hunk 3 above)
Comment (optimistyzy) [reply src/blk/pmem/PMEMDevice.h:44]: ack

File: src/blk/pmem/PMEMDevice.cc (hunk 6 above)
Comment (tchaikov) [reply src/blk/pmem/PMEMDevice.cc]: sure.

File: src/blk/pmem/PMEMDevice.cc [hunk 15]
```diff
@@ -54,6 +59,54 @@ int PMEMDevice::_lock()
...
+  FILE *sfile = fopen(spath, "r");
+  if (!sfile) {
+      return -1;
+  }
+
+  if (total_size != nullptr) {
+    rc = fscanf(sfile, "%lu", total_size);
+    if (rc < 0) {
+      rc = -1;
+   } else {
```
Comment (tchaikov) [inline src/blk/pmem/PMEMDevice.cc:101]: wrong indent.

File: src/blk/pmem/PMEMDevice.cc [hunk 16]
```diff
@@ -54,6 +59,54 @@ int PMEMDevice::_lock()
...
+     rc = 0;
+   }
```
Comment (tchaikov) [inline src/blk/pmem/PMEMDevice.cc:103]: wrong indent.

File: src/blk/pmem/PMEMDevice.cc [hunk 17]
```diff
@@ -83,7 +144,7 @@ int PMEMDevice::open(const std::string& p)
   }
 
   size_t map_len;
-  addr = (char *)pmem_map_file(path.c_str(), 0, PMEM_FILE_EXCL, O_RDWR, &map_len, NULL);
+  addr = (char *)pmem_map_file(path.c_str(), 0, devdax_device ? 0: PMEM_FILE_EXCL, O_RDWR, &map_len, NULL);
```
Comment (tchaikov) [inline src/blk/pmem/PMEMDevice.cc:147]: warp this line if you please.

**Review by tchaikov — COMMENTED** [review]
lgtm modulo some formatting nits.

File: src/blk/pmem/PMEMDevice.cc [hunk 18]
```diff
@@ -54,6 +59,54 @@ int PMEMDevice::_lock()
...
+    return -1;
+  }
+
+  // Need to check if it is a DAX device
+  char *base_name = strrchr(real_path, '/');
+  if (!base_name || strcmp("dax", base_name + 1)) {
+    return -1;
+  }
+
+  snprintf(spath, PATH_MAX, "/sys/dev/char/%d:%d/size", major(file_stat.st_rdev), minor(file_stat.st_rdev));
```
[the remark this answers is not in the document]
Comment (optimistyzy) [reply src/blk/pmem/PMEMDevice.cc:91]: ack

File: src/blk/pmem/PMEMDevice.cc [hunk 19]
```diff
@@ -54,6 +59,54 @@ int PMEMDevice::_lock()
...
+  FILE *sfile = fopen(spath, "r");
+  if (!sfile) {
+      return -1;
```
[the remark this answers is not in the document]
Comment (optimistyzy) [reply src/blk/pmem/PMEMDevice.cc:94]: ack

File: src/blk/pmem/PMEMDevice.cc (hunk 15 above)
Comment (optimistyzy) [reply src/blk/pmem/PMEMDevice.cc:101]: ack

File: src/blk/pmem/PMEMDevice.cc (hunk 16 above)
Comment (optimistyzy) [reply src/blk/pmem/PMEMDevice.cc:103]: ack

File: src/blk/pmem/PMEMDevice.cc (hunk 17 above)
Comment (optimistyzy) [reply src/blk/pmem/PMEMDevice.cc]: ack

**Review by tchaikov — APPROVED** [review]

**Comment by tchaikov** [comment]
@majianpeng the change looks good to me. do you want to have a final review to it?

**Comment by tchaikov** [comment]
jenkins test api

**Comment by majianpeng** [comment]
LGTM.

**Comment by majianpeng** [comment]
jenkins test api

**Comment by optimistyzy** [comment]
jenkins test api

**Comment by tchaikov** [comment]
building at https://shaman.ceph.com/builds/ceph/wip-pr-45609-kefu/2b3a4d668874f2c6e4fff53777fc0d7c8af796eb/

**Comment by tchaikov** [comment]
being tested at https://pulpito.ceph.com/kchai-2022-04-19_13:32:12-rados-wip-pr-45609-kefu-distro-default-smithi/

**Comment by tchaikov** [comment]
failure tracked by https://tracker.ceph.com/issues/53342

File: src/blk/pmem/PMEMDevice.cc (hunk 6 above)
Comment (tchaikov) [reply src/blk/pmem/PMEMDevice.cc]: #46122

**Pull request merged**